File: backend-go/internal/application/payments.go
package application

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v5"

	"github.com/Roy-Wanyoike/fuatilia/backend-go/internal/infra"
	"github.com/Roy-Wanyoike/fuatilia/backend-go/internal/repositories"
	"github.com/Roy-Wanyoike/fuatilia/backend-go/pkg/money"
)

// Stable codes the payments surface produces (the port of
// src/domain/payments/{intake,payment}.ts + the wire's 404s).
const (
	CodePaymentNotFound         = "HTTP_PAYMENT_NOT_FOUND"
	CodeIntakeChannelInvalid    = "INTAKE_CHANNEL_INVALID"
	CodeIntakeExternalRef       = "INTAKE_EXTERNAL_REF_REQUIRED"
	CodeIntakeIdempotencyKey    = "INTAKE_IDEMPOTENCY_KEY_REQUIRED"
	CodeIntakeDeclaredRefBlank  = "INTAKE_DECLARED_REF_BLANK"
	CodeAmountMustBePositive    = "AMOUNT_MUST_BE_POSITIVE"
	CodeDuplicateAmountMismatch = "DUPLICATE_AMOUNT_MISMATCH"
	CodeCurrencyMismatch        = "CURRENCY_MISMATCH"
	CodeConfirmedAmountMismatch = "CONFIRMED_AMOUNT_MISMATCH"
	CodePaymentTerminal         = "PAYMENT_TERMINAL"
	CodePaymentNotConfirmed     = "PAYMENT_NOT_CONFIRMED"
	CodeInvalidTransition       = "INVALID_TRANSITION"
	CodeRefundExceedsAvailable  = "REFUND_EXCEEDS_AVAILABLE"
	CodeRefundReasonRequired    = "REFUND_REASON_REQUIRED"
)

// terminalPaymentStates are docs/03's terminal edges (nothing can act on them).
var terminalPaymentStates = map[string]bool{"failed": true, "reversed": true, "refunded": true}

// confirmedFamilyStates imply the money has landed (a success callback was
// processed) — matching and refunds draw only on this family.
var confirmedFamilyStates = map[string]bool{
	"confirmed": true, "partially_allocated": true, "allocated": true,
	"unapplied": true, "partially_refunded": true, "refunded": true,
}

// IntakeCommand is the one intake funnel's input (C2B + STK converge here).
type IntakeCommand struct {
	Channel        string
	ExternalRef    string
	IdempotencyKey string
	AmountMinor    int64
	Currency       string
	CustomerID     string // "" = unknown payer (C2B paybill)
	DeclaredRefs   []string
}

// IntakeResult carries the fund truth and the R9 verdict.
type IntakeResult struct {
	Payment   repositories.PaymentRow
	Duplicate bool
}

// Intake runs the R9/C5 idempotent funnel: a duplicate
// (channel, externalRef) OR idempotencyKey replays the EXISTING payment
// (200 + duplicate:true + the duplicateCallbackObserved tripwire) and never
// creates a second Payment; a duplicate carrying different money is untrusted
// input (409). First contact inserts the payment + payment.initiated fact and
// claims the durable idempotency key — one transaction, org-scoped by the
// principal's org. A failed attempt claims nothing: the tx rolls back whole,
// so a legitimate retry always finds a free key.
func (s *Services) Intake(ctx context.Context, orgID string, cmd IntakeCommand) (IntakeResult, error) {
	channel := cmd.Channel
	if channel != "c2b" && channel != "stk" {
		return IntakeResult{}, infra.NewDomainError(CodeIntakeChannelInvalid, "unknown channel: "+channel, nil)
	}
	externalRef := trim(cmd.ExternalRef)
	if externalRef == "" {
		return IntakeResult{}, infra.NewDomainError(CodeIntakeExternalRef, "externalRef (Daraja transaction id) is required", nil)
	}
	idempotencyKey := trim(cmd.IdempotencyKey)
	if idempotencyKey == "" {
		return IntakeResult{}, infra.NewDomainError(CodeIntakeIdempotencyKey, "idempotencyKey is required (R9)", nil)
	}
	if cmd.AmountMinor <= 0 {
		return IntakeResult{}, infra.NewDomainError(CodeAmountMustBePositive, "intake amounts must be > 0", nil)
	}
	refs := make([]string, 0, len(cmd.DeclaredRefs))
	for _, raw := range cmd.DeclaredRefs {
		ref := trim(raw)
		if ref == "" {
			return IntakeResult{}, infra.NewDomainError(CodeIntakeDeclaredRefBlank, "declared references cannot be blank", nil)
		}
		if !contains(refs, ref) {
			refs = append(refs, ref)
		}
	}

	var result IntakeResult
	err := s.Stores.RunInTx(ctx, func(tx pgx.Tx) error {
		// R9 probe 1 (durable): the idempotency key may already name the
		// original outcome — hot replays never re-enter the insert path.
		if ref := s.lookupIdempotencyKey(ctx, tx, orgID, idempotencyKey); ref != "" {
			prior, err := s.stores().PaymentByID(ctx, tx, orgID, ref)
			if err == nil {
				return s.replayDuplicate(ctx, tx, orgID, prior, cmd.AmountMinor, cmd.Currency, &result)
			}
			if !errors.Is(err, repositories.ErrNotFound) {
				return err
			}
		}
		// R9 probe 2 (the lane's own uniqueness axes): unique(channel,
		// externalRef) OR unique(idempotencyKey) over the payments table.
		prior, err := s.findPriorPayment(ctx, tx, orgID, channel, externalRef, idempotencyKey)
		if err != nil {
			return err
		}
		if prior != nil {
			if err := s.assertDuplicateMoney(*prior, cmd.AmountMinor, cmd.Currency); err != nil {
				return err
			}
			// Bind this key to the original outcome too, so future replays
			// carrying the same key find it through the durable registry.
			if _, _, err := s.claimIdempotencyKey(ctx, tx, orgID, idempotencyKey, prior.ID); err != nil {
				return err
			}
			if err := s.appendDuplicateTripwire(ctx, tx, orgID, *prior); err != nil {
				return err
			}
			s.rememberReplay(orgID, idempotencyKey, prior.ID)
			result = IntakeResult{Payment: *prior, Duplicate: true}
			return nil
		}

		// First contact: claim the key, insert the payment, append the fact.
		payment := repositories.PaymentRow{
			ID:             s.IDs(),
			OrgID:          orgID,
			CustomerID:     strPtrOf(cmd.CustomerID),
			Channel:        channel,
			ExternalRef:    externalRef,
			IdempotencyKey: idempotencyKey,
			State:          "initiated",
			Currency:       cmd.Currency,
			RequestedMinor: cmd.AmountMinor,
			UnappliedMinor: zeroPtr(),
			DeclaredRefs:   refs,
			InitiatedAt:    s.Clock.Now(),
		}
		if _, _, err := s.claimIdempotencyKey(ctx, tx, orgID, idempotencyKey, payment.ID); err != nil {
			return err
		}
		if err := s.stores().InsertPayment(ctx, tx, payment); err != nil {
			if repositories.UniqueViolation(err) {
				// A concurrent callback won the race. This transaction is
				// aborted (PostgreSQL semantics); it rolls back whole — the
				// durable key claim included — and the replay is resolved on
				// a fresh read below against the winner's committed state.
				return errRetryReplay
			}
			return err
		}
		if err := s.appendOutbox(ctx, tx, orgID, "payment.initiated", payment.ID, map[string]any{
			"paymentId":      payment.ID,
			"channel":        channel,
			"requestedMinor": cmd.AmountMinor,
		}); err != nil {
			return err
		}
		s.rememberReplay(orgID, idempotencyKey, payment.ID)
		result = IntakeResult{Payment: payment, Duplicate: false}
		return nil
	})
	if errors.Is(err, errRetryReplay) {
		// The winner committed between our probe and our insert; replay it.
		return s.replayAfterRace(ctx, orgID, channel, externalRef, idempotencyKey, cmd.AmountMinor, cmd.Currency)
	}
	if err != nil {
		return IntakeResult{}, err
	}
	return result, nil
}

// errRetryReplay signals "this tx lost an insert race — replay the winner on
// a fresh transaction". Internal only; never leaves the service.
var errRetryReplay = errors.New("application: intake insert lost a race — replay the committed winner")

// replayAfterRace resolves the concurrent winner's payment in a fresh
// transaction (the losing tx rolled back, so nothing it claimed survives).
func (s *Services) replayAfterRace(ctx context.Context, orgID, channel, externalRef, idempotencyKey string, amountMinor int64, currency string) (IntakeResult, error) {
	var result IntakeResult
	err := s.Stores.RunInTx(ctx, func(tx pgx.Tx) error {
		prior, err := s.findPriorPayment(ctx, tx, orgID, channel, externalRef, idempotencyKey)
		if err != nil {
			return err
		}
		if prior == nil {
			return infra.NewDomainError(infra.CodeInternal, "intake race resolved without a committed payment", nil)
		}
		if err := s.assertDuplicateMoney(*prior, amountMinor, currency); err != nil {
			return err
		}
		if _, _, err := s.claimIdempotencyKey(ctx, tx, orgID, idempotencyKey, prior.ID); err != nil {
			return err
		}
		if err := s.appendDuplicateTripwire(ctx, tx, orgID, *prior); err != nil {
			return err
		}
		s.rememberReplay(orgID, idempotencyKey, prior.ID)
		result = IntakeResult{Payment: *prior, Duplicate: true}
		return nil
	})
	if err != nil {
		return IntakeResult{}, err
	}
	return result, nil
}

// findPriorPayment is the R9 duplicate probe: same (channel, externalRef) OR
// same idempotencyKey replays the same logical payment.
func (s *Services) findPriorPayment(ctx context.Context, q repositories.Querier, orgID, channel, externalRef, idempotencyKey string) (*repositories.PaymentRow, error) {
	prior, err := s.stores().PaymentByExternalRef(ctx, q, orgID, channel, externalRef)
	if err == nil {
		return &prior, nil
	}
	if !errors.Is(err, repositories.ErrNotFound) {
		return nil, err
	}
	prior, err = s.stores().PaymentByIdempotencyKey(ctx, q, orgID, idempotencyKey)
	if err == nil {
		return &prior, nil
	}
	if errors.Is(err, repositories.ErrNotFound) {
		return nil, nil
	}
	return nil, err
}

// replayDuplicate runs the duplicate path against a payment resolved by the
// durable key: money parity first (a tampered replay is untrusted input,
// never a benign retry), then the tripwire + the ORIGINAL result.
func (s *Services) replayDuplicate(ctx context.Context, tx repositories.Querier, orgID string, prior repositories.PaymentRow, amountMinor int64, currency string, result *IntakeResult) error {
	if err := s.assertDuplicateMoney(prior, amountMinor, currency); err != nil {
		return err
	}
	if err := s.appendDuplicateTripwire(ctx, tx, orgID, prior); err != nil {
		return err
	}
	result.Payment = prior
	result.Duplicate = true
	return nil
}

// appendDuplicateTripwire records the C5 ops fact
// (payments.duplicateCallbackObserved) beside the replayed payment — the
// at-least-once channel's observable heartbeat (E15's payload shape).
func (s *Services) appendDuplicateTripwire(ctx context.Context, tx repositories.Querier, orgID string, prior repositories.PaymentRow) error {
	return s.appendOutbox(ctx, tx, orgID, "payments.duplicateCallbackObserved", prior.ID, map[string]any{
		"paymentId":   prior.ID,
		"externalRef": prior.ExternalRef,
		"seenAt":      repositories.ISO(s.Clock.Now()),
	})
}

// assertDuplicateMoney mirrors intake.ts: same transaction id must mean the
// same money — a different currency (R10) or amount is untrusted input.
func (s *Services) assertDuplicateMoney(prior repositories.PaymentRow, amountMinor int64, currency string) error {
	if prior.Currency != currency {
		return infra.NewDomainError(CodeCurrencyMismatch,
			"payment "+prior.ID+" is "+prior.Currency+"; duplicate callback arrived as "+currency+" (R10)", nil)
	}
	if prior.RequestedMinor != amountMinor {
		priorAmount, _ := moneyString(prior.RequestedMinor, prior.Currency)
		callbackAmount, _ := moneyString(amountMinor, currency)
		return infra.NewDomainError(CodeDuplicateAmountMismatch,
			"duplicate callback for "+prior.ExternalRef+" carries "+callbackAmount+" but the payment was initiated for "+priorAmount, nil)
	}
	return nil
}

// ConfirmResult is the success callback's verdict.
type ConfirmResult struct {
	Payment          repositories.PaymentRow
	AlreadyConfirmed bool
}

// Confirm processes the Daraja success callback idempotently: confirmed is
// set exactly ONCE; a same-amount replay is a 200 no-op, a different amount
// is CONFIRMED_AMOUNT_MISMATCH. The state change, the payment.confirmed fact
// AND the balanced ledger entry (R4: Σdebits == Σcredits) commit in ONE
// transaction — the journal ref is the replay key, so a re-confirm can never
// double-post.
func (s *Services) Confirm(ctx context.Context, orgID, paymentID string, amountMinor int64, currency string) (ConfirmResult, error) {
	if amountMinor <= 0 {
		return ConfirmResult{}, infra.NewDomainError(CodeAmountMustBePositive, "confirmation amounts must be > 0", nil)
	}
	var result ConfirmResult
	err := s.Stores.RunInTx(ctx, func(tx pgx.Tx) error {
		payment, err := s.stores().PaymentByID(ctx, tx, orgID, paymentID)
		if err != nil {
			if errors.Is(err, repositories.ErrNotFound) {
				return infra.NewDomainError(CodePaymentNotFound, "payment "+paymentID+" does not exist", nil)
			}
			return err
		}
		if payment.Currency != currency {
			return infra.NewDomainError(CodeCurrencyMismatch,
				"cannot confirm "+currency+" against a "+payment.Currency+" payment (R10)", nil)
		}
		if terminalPaymentStates[payment.State] {
			return infra.NewDomainError(CodePaymentTerminal,
				"payment "+payment.ID+" is "+payment.State+" (terminal); confirmation is not allowed", nil)
		}
		if payment.State == "confirmed" {
			if payment.ConfirmedMinor == nil {
				return infra.NewDomainError(infra.CodeInternal, "payment "+payment.ID+" is confirmed but has no confirmed amount", nil)
			}
			if *payment.ConfirmedMinor != amountMinor {
				return infra.NewDomainError(CodeConfirmedAmountMismatch,
					"payment "+payment.ID+" already confirmed for "+moneyStringOf(*payment.ConfirmedMinor, payment.Currency)+
						", cannot re-confirm for "+moneyStringOf(amountMinor, payment.Currency), nil)
			}
			result = ConfirmResult{Payment: payment, AlreadyConfirmed: true}
			return nil
		}
		if payment.State == "initiated" {
			// The callback may race ahead of the platform's
			// awaiting-confirmation step — advance through the lane's own
			// transition first.
			if err := s.stores().StagePaymentForConfirmation(ctx, tx, orgID, payment.ID); err != nil {
				return err
			}
			payment.State = "pending_confirmation"
		}
		if payment.State != "pending_confirmation" {
			return infra.NewDomainError(CodeInvalidTransition,
				"confirmation applies to pending_confirmation payments, got "+payment.State, nil)
		}
		now := s.Clock.Now()
		if err := s.stores().ConfirmPayment(ctx, tx, orgID, payment.ID, amountMinor, now); err != nil {
			return err
		}
		if err := s.appendOutbox(ctx, tx, orgID, "payment.confirmed", payment.ID, map[string]any{
			"paymentId":      payment.ID,
			"confirmedMinor": amountMinor,
			"externalRef":    payment.ExternalRef,
			"confirmedAt":    repositories.ISO(now),
		}); err != nil {
			return err
		}
		if err := s.stores().EnsureConfirmationLedgerSeed(ctx, tx, orgID, currency); err != nil {
			return err
		}
		if err := s.stores().PostConfirmationEntry(ctx, tx, orgID, payment.ID, payment.ExternalRef, currency, amountMinor, now); err != nil {
			return err
		}
		payment.State = "confirmed"
		payment.ConfirmedMinor = &amountMinor
		unapplied := amountMinor
		payment.UnappliedMinor = &unapplied
		payment.ConfirmedAt = &now
		result = ConfirmResult{Payment: payment, AlreadyConfirmed: false}
		return nil
	})
	if err != nil {
		return ConfirmResult{}, err
	}
	return result, nil
}

// RefundResult carries the payment view after the reservation appended.
type RefundResult struct {
	Payment repositories.PaymentRow
}

// RefundReservation records an append-only refund reservation (R6): refunds
// draw ONLY on funds not already allocated/refunded — over-draw refuses with
// REFUND_EXCEEDS_AVAILABLE (422) and the tx rolls back with zero rows.
// State-neutral by design: the Refunded/PartiallyRefunded edges belong to the
// adjustments lane's Refund aggregate; the reservation row keeps the
// payment-side ceiling honest (the DDL re-proves it at COMMIT).
func (s *Services) RefundReservation(ctx context.Context, orgID, paymentID, requestedBy string, amountMinor int64, currency, reason string) (RefundResult, error) {
	why := trim(reason)
	if why == "" {
		return RefundResult{}, infra.NewDomainError(CodeRefundReasonRequired, "a refund reservation requires a reason", nil)
	}
	if amountMinor <= 0 {
		return RefundResult{}, infra.NewDomainError(CodeAmountMustBePositive, "refund amounts must be > 0", nil)
	}
	var result RefundResult
	err := s.Stores.RunInTx(ctx, func(tx pgx.Tx) error {
		payment, err := s.stores().PaymentByID(ctx, tx, orgID, paymentID)
		if err != nil {
			if errors.Is(err, repositories.ErrNotFound) {
				return infra.NewDomainError(CodePaymentNotFound, "payment "+paymentID+" does not exist", nil)
			}
			return err
		}
		if payment.Currency != currency {
			return infra.NewDomainError(CodeCurrencyMismatch,
				"cannot refund "+currency+" against a "+payment.Currency+" payment (R10)", nil)
		}
		if terminalPaymentStates[payment.State] {
			return infra.NewDomainError(CodePaymentTerminal,
				"payment "+payment.ID+" is "+payment.State+" (terminal); refund is not allowed", nil)
		}
		if !confirmedFamilyStates[payment.State] {
			return infra.NewDomainError(CodePaymentNotConfirmed,
				"refunds draw on confirmed funds; payment "+payment.ID+" is "+payment.State+" (R6)", nil)
		}
		if payment.ConfirmedMinor == nil {
			return infra.NewDomainError(infra.CodeInternal, "payment "+payment.ID+" is confirmed but has no confirmed amount", nil)
		}
		allocations, refunds, err := s.stores().CommittedAgainstPayment(ctx, tx, orgID, payment.ID)
		if err != nil {
			return err
		}
		committed := allocations + refunds + amountMinor
		if committed > *payment.ConfirmedMinor {
			return infra.NewDomainError(CodeRefundExceedsAvailable, exceedsMessage(committed, *payment.ConfirmedMinor, currency), nil)
		}
		row := repositories.RefundRow{
			ID:        s.IDs(),
			OrgID:     orgID,
			PaymentID: payment.ID,
			Amount:    amountMinor,
			Currency:  currency,
			Reason:    why,
			CreatedAt: s.Clock.Now(),
		}
		if err := s.stores().InsertRefundReservation(ctx, tx, row, trim(requestedBy)); err != nil {
			// The deferrable R6 trigger re-proves the ceiling at COMMIT —
			// its refusal is the same domain outcome as the pre-check.
			if repositories.CheckViolation(err) {
				return infra.NewDomainError(CodeRefundExceedsAvailable, exceedsMessage(committed, *payment.ConfirmedMinor, currency), nil)
			}
			return err
		}
		result = RefundResult{Payment: payment}
		return nil
	})
	if err != nil {
		return RefundResult{}, err
	}
	return result, nil
}

// exceedsMessage renders the R6 refusal exactly as the TS lane words it.
func exceedsMessage(committed, confirmed int64, currency string) string {
	committedStr, _ := moneyString(committed, currency)
	confirmedStr, _ := moneyString(confirmed, currency)
	return "Σ allocations+refunds " + committedStr + " would exceed confirmed " + confirmedStr + " (R6)"
}

// PaymentWithRows is the payment aggregate's full read shape.
type PaymentWithRows struct {
	Payment     repositories.PaymentRow
	Allocations []repositories.AllocationRow
	Refunds     []repositories.RefundRow
}

// GetPayment loads one org-scoped payment with its posting rows.
func (s *Services) GetPayment(ctx context.Context, q repositories.Querier, orgID, paymentID string) (PaymentWithRows, error) {
	payment, err := s.stores().PaymentByID(ctx, q, orgID, paymentID)
	if err != nil {
		if errors.Is(err, repositories.ErrNotFound) {
			return PaymentWithRows{}, infra.NewDomainError(CodePaymentNotFound, "payment "+paymentID+" does not exist", nil)
		}
		return PaymentWithRows{}, err
	}
	allocations, err := s.stores().AllocationsForPayment(ctx, q, orgID, payment.ID)
	if err != nil {
		return PaymentWithRows{}, err
	}
	refunds, err := s.stores().RefundsForPayment(ctx, q, orgID, payment.ID)
	if err != nil {
		return PaymentWithRows{}, err
	}
	return PaymentWithRows{Payment: payment, Allocations: allocations, Refunds: refunds}, nil
}

// ListPayments is the org-scoped paginated read model.
func (s *Services) ListPayments(ctx context.Context, q repositories.Querier, orgID, sortCol, order string, limit, offset int) ([]repositories.PaymentRow, int, error) {
	return s.stores().PaymentsByOrg(ctx, q, orgID, sortCol, order, limit, offset)
}

func (s *Services) stores() *repositories.Stores { return s.Stores }

func trim(raw string) string {
	start, end := 0, len(raw)
	for start < end && (raw[start] == ' ' || raw[start] == '\t' || raw[start] == '\n' || raw[start] == '\r') {
		start++
	}
	for end > start && (raw[end-1] == ' ' || raw[end-1] == '\t' || raw[end-1] == '\n' || raw[end-1] == '\r') {
		end--
	}
	return raw[start:end]
}

func contains(list []string, v string) bool {
	for _, item := range list {
		if item == v {
			return true
		}
	}
	return false
}

func zeroPtr() *int64 {
	zero := int64(0)
	return &zero
}

// moneyStringOf renders via pkg/money — the TS Money.toString parity
// ("whole.cents CUR") the mismatch messages rely on. The plain literal is
// the honest fallback for a currency outside the closed set (which the DDL
// otherwise already refuses at the column).
func moneyStringOf(minor int64, currency string) string {
	out, err := moneyString(minor, currency)
	if err != nil {
		return itoa(minor) + " " + currency
	}
	return out
}

func moneyString(minor int64, currency string) (string, error) {
	m, err := money.New(minor, money.Currency(currency))
	if err != nil {
		return itoa(minor) + " " + currency, nil
	}
	return m.String(), nil
}

File: src/adapters/http/routes/payments.ts
/**
 * `/v1/payments/*` — the fund-truth surface over the payments lane (issue
 * #60): the ONE intake funnel, lookup/read-model and the refund lifecycle.
 *
 * The route registration is a TABLE of `{ method, pattern, permission,
 * handler }` rows, in the style of `routes/auth.ts`; permissions come from
 * the closed vocabulary:
 *
 *   - `payments:intake` — the Daraja at-least-once funnel: intake (C2B
 *     callbacks + STK push results converge here, R9/C5) and the success
 *     callback (confirmation). Replays are NORMAL and idempotent:
 *     duplicate intake → 200 with the EXISTING payment and the lane's
 *     `payments.duplicateCallbackObserved` tripwire recorded; a re-confirmed
 *     success callback with the SAME amount → 200 `alreadyConfirmed`, with a
 *     DIFFERENT amount → 409 `CONFIRMED_AMOUNT_MISMATCH`.
 *   - `payments:read` — lookup / list (read model).
 *   - `payments:refund` — refund reservations (R6: refunds draw only on
 *     funds not already allocated/refunded; over-draw → 422
 *     `REFUND_EXCEEDS_AVAILABLE`).
 *
 * Handlers are wire→lane adapters ONLY: validate body shape
 * (`HTTP_BODY_INVALID`), look up the aggregate (`HTTP_PAYMENT_NOT_FOUND`),
 * call the lane's pure functions with the injected clock/ids (money math is
 * the lane's — never re-implemented here), persist through the injected
 * ResourceStore, record the lane's events, project serializable views.
 * Refusal-as-value decisions map per the kernel's status table (409/422).
 *
 * Confirmation orchestration note: the wire route represents the Daraja
 * success callback. A callback may race ahead of the platform's
 * "awaiting confirmation" step, so the adapter advances an `initiated`
 * payment through the lane's own `awaitConfirmation` transition before
 * calling `confirmPayment` — lane transitions, no re-implementation.
 *
 * Org scoping: the payment aggregate carries no orgId (lane value) — see
 * runtime/resources.ts for the reference-store scoping note.
 */
import type { Uuid } from '../../../domain/shared/ids';
import { DomainError } from '../../../domain/shared/errors';
import { CURRENCIES, Money, type Currency } from '../../../domain/shared/money';
import { intakePayment } from '../../../domain/payments/intake';
import {
  awaitConfirmation,
  confirmPayment,
  recordRefundReservation,
  unappliedMinorOf,
  type Payment,
} from '../../../domain/payments/payment';
import { HTTP_BODY_INVALID, HTTP_PAYMENT_NOT_FOUND, HTTP_QUERY_INVALID } from '../kernel/errors';
import type { RequestContext, RouteRecord } from '../kernel/types';
import { paginatedMeta, parsePagination, parseSorting } from '../pagination';
import { toStoredEvent, type ResourceRouteDeps } from '../runtime/resources';

// --- body field guards (wire-shape validation only — the domain re-validates values) ----

const bodyObject = (body: unknown): Record<string, unknown> => {
  if (typeof body !== 'object' || body === null || Array.isArray(body)) {
    throw new DomainError(HTTP_BODY_INVALID, 'request body must be a JSON object');
  }
  return body as Record<string, unknown>;
};

const stringField = (body: Record<string, unknown>, name: string): string => {
  const value = body[name];
  if (typeof value !== 'string' || value.trim() === '') {
    throw new DomainError(HTTP_BODY_INVALID, `field '${name}' must be a non-empty string`);
  }
  return value.trim();
};

const uuidField = (body: Record<string, unknown>, name: string): Uuid => {
  const raw = stringField(body, name);
  try {
    return uuidOf(raw);
  } catch {
    throw new DomainError(HTTP_BODY_INVALID, `field '${name}' must be a UUID`);
  }
};

const optionalUuidField = (body: Record<string, unknown>, name: string): Uuid | undefined => {
  if (body[name] === undefined) return undefined;
  return uuidField(body, name);
};

const optionalStringArrayField = (body: Record<string, unknown>, name: string): readonly string[] | undefined => {
  const raw = body[name];
  if (raw === undefined) return undefined;
  if (!Array.isArray(raw) || raw.some((entry) => typeof entry !== 'string')) {
    throw new DomainError(HTTP_BODY_INVALID, `field '${name}' must be an array of strings`);
  }
  return raw as string[];
};

/** Money on the wire: `{ minor, currency }` — minor units as a safe positive integer (R10). */
const moneyField = (body: Record<string, unknown>, name: string): Money => {
  const raw = body[name];
  if (typeof raw !== 'object' || raw === null || Array.isArray(raw)) {
    throw new DomainError(HTTP_BODY_INVALID, `field '${name}' must be an object { minor, currency }`);
  }
  const shape = raw as Record<string, unknown>;
  const minor = shape['minor'];
  const currency = shape['currency'];
  if (typeof minor !== 'number' || !Number.isSafeInteger(minor) || minor <= 0) {
    throw new DomainError(HTTP_BODY_INVALID, `field '${name}.minor' must be a positive integer (minor units)`);
  }
  if (typeof currency !== 'string' || !(CURRENCIES as readonly string[]).includes(currency)) {
    throw new DomainError(HTTP_BODY_INVALID, `field '${name}.currency' must be one of: ${CURRENCIES.join(', ')}`);
  }
  return Money.ofMinor(minor, currency as Currency);
};

const uuidOf = (raw: string): Uuid => {
  if (!/^[0-9a-fA-F-]{36}$/.test(raw)) throw new Error(`invalid uuid: ${raw}`);
  return raw as Uuid;
};

// --- serializable views -----------------------------------------------------------------

const jsonMoney = (amount: Money): { minor: number; currency: Currency } => ({
  minor: Number(amount.amount),
  currency: amount.currency,
});

const isoOrNull = (at: Date | undefined | null): string | null => (at ? at.toISOString() : null);

export const paymentView = (payment: Payment) => ({
  id: payment.id,
  channel: payment.channel,
  externalRef: payment.externalRef,
  idempotencyKey: payment.idempotencyKey,
  customerId: payment.customerId ?? null,
  state: payment.state,
  currency: payment.currency,
  requested: jsonMoney(payment.requestedMinor),
  confirmed: payment.confirmedMinor ? jsonMoney(payment.confirmedMinor) : null,
  // Derivable unapplied balance — the lane's own ceiling math (R2/R6), not the handler's.
  unapplied: jsonMoney(unappliedMinorOf(payment)),
  declaredRefs: [...payment.declaredRefs],
  allocations: payment.allocations.map((row) => ({
    id: row.id,
    receivableId: row.receivableId,
    amount: jsonMoney(row.amount),
    recordedAt: row.recordedAt.toISOString(),
  })),
  refunds: payment.refunds.map((row) => ({
    id: row.id,
    amount: jsonMoney(row.amount),
    reason: row.reason,
    recordedAt: row.recordedAt.toISOString(),
  })),
  initiatedAt: payment.initiatedAt.toISOString(),
  confirmedAt: isoOrNull(payment.confirmedAt),
  failedAt: isoOrNull(payment.failedAt),
  failureCode: payment.failureCode ?? null,
  reversedAt: isoOrNull(payment.reversedAt),
  reversalReason: payment.reversalReason ?? null,
});

// --- the route table -----------------------------------------------------------------------

const SORTABLE = ['id', 'state', 'initiatedAt'] as const;

/** The opaque reference cursor is the offset into the deterministic order. */
const decodeCursor = (cursor: string): number => {
  const offset = Number(cursor);
  if (!Number.isInteger(offset) || offset < 0) {
    throw new DomainError(HTTP_QUERY_INVALID, `query parameter 'cursor' is not a valid page cursor`);
  }
  return offset;
};

export function paymentsRoutes(deps: ResourceRouteDeps): RouteRecord[] {
  const { store, clock, idGen } = deps;

  const intakeRoute: RouteRecord = {
    method: 'POST',
    pattern: '/v1/payments/intake',
    permission: 'payments:intake',
    handler: (ctx) => {
      const body = bodyObject(ctx.body);
      const channel = body['channel'];
      if (channel !== 'c2b' && channel !== 'stk') {
        throw new DomainError(HTTP_BODY_INVALID, "field 'channel' must be 'c2b' or 'stk'");
      }
      const result = intakePayment(
        {
          channel,
          externalRef: stringField(body, 'externalRef'),
          idempotencyKey: stringField(body, 'idempotencyKey'),
          amount: moneyField(body, 'amount'),
          customerId: optionalUuidField(body, 'customerId'),
          declaredRefs: optionalStringArrayField(body, 'declaredRefs'),
          paymentId: idGen() as Uuid,
        },
        { clock, existing: store.payments() },
      );
      store.savePayment(result.payment);
      for (const event of result.events) store.record(toStoredEvent(event));
      return {
        // R9/C5 replay semantics: a duplicate is the SAME logical payment —
        // 200 with the existing row, never a second Payment.
        status: result.duplicate ? 200 : 201,
        data: { payment: paymentView(result.payment), duplicate: result.duplicate },
      };
    },
  };

  const getPaymentRoute: RouteRecord = {
    method: 'GET',
    pattern: '/v1/payments/:paymentId',
    permission: 'payments:read',
    handler: (ctx) => {
      const payment = store.payments().find((p) => p.id === ctx.params['paymentId']);
      if (!payment) {
        throw new DomainError(HTTP_PAYMENT_NOT_FOUND, `payment ${ctx.params['paymentId']} does not exist`);
      }
      return { status: 200, data: { payment: paymentView(payment) } };
    },
  };

  const listPaymentsRoute: RouteRecord = {
    method: 'GET',
    pattern: '/v1/payments',
    permission: 'payments:read',
    handler: (ctx) => {
      const { limit, cursor } = parsePagination(ctx.query);
      const { field, order } = parseSorting(ctx.query, SORTABLE);
      const offset = cursor === null ? 0 : decodeCursor(cursor);
      const views = store.payments().map((payment) => paymentView(payment) as unknown as Record<string, unknown>);
      const ordered =
        field === null ? views : [...views].sort((a, b) => {
          const cmp = String(a[field]).localeCompare(String(b[field]));
          return order === 'asc' ? cmp : -cmp;
        });
      const page = ordered.slice(offset, offset + limit);
      return {
        status: 200,
        data: { payments: page },
        meta: paginatedMeta(offset + limit < ordered.length ? String(offset + limit) : null, ordered.length),
      };
    },
  };

  const confirmRoute: RouteRecord = {
    method: 'POST',
    pattern: '/v1/payments/:paymentId/confirmations',
    permission: 'payments:intake',
    handler: (ctx) => {
      const payment = store.payments().find((p) => p.id === ctx.params['paymentId']);
      if (!payment) {
        throw new DomainError(HTTP_PAYMENT_NOT_FOUND, `payment ${ctx.params['paymentId']} does not exist`);
      }
      const amount = moneyField(bodyObject(ctx.body), 'amount');
      if (payment.state === 'confirmed') {
        // Replay of the same success callback: the lane validates the amount
        // (a DIFFERENT amount throws CONFIRMED_AMOUNT_MISMATCH → 409) and
        // answers a no-op — confirmedMinor is set exactly ONCE.
        confirmPayment(payment, amount, clock);
        return { status: 200, data: { payment: paymentView(payment), alreadyConfirmed: true } };
      }
      const staged = payment.state === 'initiated' ? awaitConfirmation(payment).payment : payment;
      const { payment: confirmed, events } = confirmPayment(staged, amount, clock);
      store.savePayment(confirmed);
      for (const event of events) store.record(toStoredEvent(event));
      return { status: 201, data: { payment: paymentView(confirmed), alreadyConfirmed: false } };
    },
  };

  const refundRoute: RouteRecord = {
    method: 'POST',
    pattern: '/v1/payments/:paymentId/refund-reservations',
    permission: 'payments:refund',
    handler: (ctx) => {
      const payment = store.payments().find((p) => p.id === ctx.params['paymentId']);
      if (!payment) {
        throw new DomainError(HTTP_PAYMENT_NOT_FOUND, `payment ${ctx.params['paymentId']} does not exist`);
      }
      const body = bodyObject(ctx.body);
      const result = recordRefundReservation(
        payment,
        { amount: moneyField(body, 'amount'), reason: stringField(body, 'reason') },
        clock,
      );
      store.savePayment(result.payment);
      // The lane emits no event here: the Refunded/PartiallyRefunded edges
      // belong to the adjustments lane's Refund aggregate (issue #4); the
      // reservation row keeps the payment-side R6 ceiling honest.
      return { status: 201, data: { payment: paymentView(result.payment) } };
    },
  };

  return [intakeRoute, getPaymentRoute, listPaymentsRoute, confirmRoute, refundRoute];
}
